pkg/controller/bastion/actuator_reconcile.go

```diff
@@ -0,0 +1,327 @@
+}
+
+// getMachineEndpoints function returns the bastion endpoints of a running
+// machine. It first validates that the machine is in running state, then
+// extracts the private and public IP of the machine's network interface, and
+// finally converts the IPs to their respective ingress addresses.
+func getMachineEndpoints(machine *computev1alpha1.Machine) (*bastionEndpoints, error) {
+	if machine == nil {
+		return nil, fmt.Errorf("machine can't be nil")
+	}
+
+	if machine.Status.State != computev1alpha1.MachineStateRunning {
+		return nil, fmt.Errorf("machine not running, status: %s", machine.Status.State)
+	}
+
+	endpoints := &bastionEndpoints{}
+
+	if len(machine.Status.NetworkInterfaces) == 0 {
+		return nil, fmt.Errorf("no network interface found for machine: %s", machine.Name)
+	}
+
+	if len(machine.Status.NetworkInterfaces[0].IPs) == 0 {
+		return nil, fmt.Errorf("no private ip found for network interface for machine: %s", machine.Name)
+	}
+
+	privateIP := machine.Status.NetworkInterfaces[0].IPs[0].String()
+	if ingress := addressToIngress(&machine.Name, &privateIP); ingress != nil {
+		endpoints.private = ingress
+	}
+
+	if machine.Status.NetworkInterfaces[0].VirtualIP == nil {
+		return nil, fmt.Errorf("no virtual ip found for network interface for machine: %s", machine.Name)
+	}
+
+	virtualIP := machine.Status.NetworkInterfaces[0].VirtualIP.String()
+	if ingress := addressToIngress(&machine.Name, &virtualIP); ingress != nil {
+		endpoints.public = ingress
+	}
+
+	return endpoints, nil
+}
+
+// bastionConfigCheck checks whether a bastion configuration is valid.
+func bastionConfigCheck(bastionConfig *controllerconfig.BastionConfig) error {
+	if bastionConfig == nil {
+		return fmt.Errorf("bastionConfig must not be empty")
+	}
+
+	if bastionConfig.MachineClassName == "" {
+		return fmt.Errorf("bastion not supported as no machine class is configured for the bastion host machine")
+	}
+
+	if bastionConfig.Image == "" {
+		return fmt.Errorf("bastion not supported as no Image is configured for the bastion host machine")
+	}
+	return nil
+}
+
+// applyMachine applies the configuration to create or update the bastion host machine
+// and the ignition secret used for provisioning the machine. It first sets the owner
+// reference for the ignition secret to the bastion host machine, to ensure that the secret
+// is garbage collected when the machine is deleted.
+func applyMachine(ctx context.Context, onmetalClient client.Client, ignitionSecret *corev1.Secret, bastionHost *computev1alpha1.Machine) (*computev1alpha1.Machine, error) {
+
+	_, err := controllerutil.CreateOrPatch(ctx, onmetalClient, bastionHost, nil)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create or patch bastion host machine %s: %w", client.ObjectKeyFromObject(bastionHost), err)
+	}
+
+	//TODO: Even after its owner is set as the Bastion machine, ignitionSecret does not get garbage collected when deleting Bastion.
```

Comment:
Does this issue still occur?

Reply:
yes, able to set owner reference correctly, but still ignition secret is not getting garbage collected so we have to explicitly remove it.